### PR TITLE
[RFC] Pareto front visualization to visualize study progress with color scales

### PR DESCRIPTION
**Pull request by hvy**
## Motivation

The Pareto front visualizations (2d, 3d in `plot_pareto_front`) visualize trials with scatter plots, distinguishing between trial in the front and ones that are not using two different colors. Similar to [`plot_slice`](https://optuna.readthedocs.io/en/stable/reference/visualization/generated/optuna.visualization.plot_slice.html#optuna.visualization.plot_slice), they could additionally show the study progress using color scales in Plotly. This information is e.g. useful to study what tendencies the search algorithms have and how the search changes during the course of the study.

### `master`

https://optuna.readthedocs.io/en/stable/reference/visualization/generated/optuna.visualization.plot_pareto_front.html#optuna.visualization.plot_pareto_front

### This PR

newplot (37)

newplot (36)

### CODE REVIEW

**Comment by hvy** [comment]
@sile , @nzw0301 could you possibly review this PR? I'll reassign it if you're busy.

**Review by nzw0301 — APPROVED** [review]
It looks great! I locally checked the plot with`include_dominated_trials=False`:

newplot (1)

File: optuna/visualization/_pareto_front.py [hunk 1]
```diff
@@ -138,15 +138,33 @@ def _get_pareto_front_2d(
             text=[_make_hovertext(t) for t in trials[len(study.best_trials) :]],
             mode="markers",
             hovertemplate="%{text}<extra>Trial</extra>",
-            name="Trial",
+            marker={
+                "line": {"width": 0.5, "color": "Grey"},
+                "color": [t.number for t in trials[len(study.best_trials) :]],
```
Comment (sile) [inline optuna/visualization/_pareto_front.py:143]: IIUC, `study.best_trials` could be a somewhat heavy call when there are too many trials, so how about storing the best trials into a variable at the head of this function and reusing that here (and other places)?

**Review by sile — CHANGES_REQUESTED** [review]

File: optuna/visualization/_pareto_front.py (hunk 1 above)
Comment (hvy) [reply optuna/visualization/_pareto_front.py]: Good catch, thanks. It should be fixed in the latest commit.

**Comment by hvy** [comment]
PTAL.

File: optuna/visualization/_pareto_front.py (hunk 1 above)
Comment (sile) [reply optuna/visualization/_pareto_front.py]: Thanks!

**Review by sile — APPROVED** [review]
LGTM!

**Pull request merged**